simfold/src/common/common.cpp

```diff
@@ -649,87 +649,130 @@ PARAMTYPE IL_penalty_by_size_2D (int size1, int size2)
+                    printf("Error in simfold/src/common/common.cpp creation of size_penalties\n");
+                    exit(-1);
+            }
+        }
+        else if (parsi_length == PARSI || parsi_length == ZL)
+        {
+            switch(type) {
+                case 0:
+                    end = MAXLOOP_H_PARSI; break;
+                case 1:
+                    end = MAXLOOP_B_PARSI; break;
+                case 2:
+                    end = MAXLOOP_I_PARSI; break;
+                default:
+                    printf("Error in simfold/src/common/common.cpp creation of size_penalties\n");
+                    exit(-1);
+            }
+        }
+        else if (parsi_length == LAVISH)
+        {
+            switch(type) {
+                case 0:
+                    end = MAXLOOP_H_LAVISH; break;
+                case 1:
+                    end = MAXLOOP_B_LAVISH; break;
+                case 2:
+                    end = MAXLOOP_I_LAVISH; break;
+                default:
+                    printf("Error in simfold/src/common/common.cpp creation of size_penalties\n");
+                    exit(-1);
+            }
+        }
+
+        arr[type].resize(max_size);
+        for (int size=0; size < max_size; size++) {
+            PARAMTYPE penalty30, penalty;
+            double logval;
+
+            // the penalties for size <= MAXLOOP _H, _B, _I should be read from the file "loop"
+            //if (size <= MAXLOOP)
+            if (type == 0 && size <= end)
+            {
+                penalty = hairpin_penalty_by_size[size];
+            }
+            else if (type == 1 && size <= end)
+            {
+                penalty = bulge_penalty_by_size[size];
+            }
+            else if (type == 2 && size <= end)
+            {
+                penalty = internal_penalty_by_size[size];
+            }
+            else {
+                //return 50.0;
+                // size > MAXLOOP _H, _B, _I
+                if (type == 0)
+                {
+                    penalty30 = hairpin_penalty_by_size[end];
+                    logval = log (1.0*size/end);
+                }
+                else if (type == 1)
+                {
+                    penalty30 = bulge_penalty_by_size[end];
+                    logval = log (1.0*size/end);
+                }
+                else if (type == 2)
+                {
+                    penalty30 = internal_penalty_by_size[end];
+                    logval = log (1.0*size/end);
+                }
+                else
+                {
+                    printf ("ERROR! type is not valid, ABORT!\n");
+                    exit(1);
+                }
+
+                penalty = (PARAMTYPE) (penalty30 + 100.0*misc.param_greater30 * logval);
+            }
+
+            arr[type][size] = penalty;
+        }
+    }
+}
+
+size_penalties_class::~size_penalties_class() {
+
+}
+
+PARAMTYPE size_penalties_class::get_size_penalty(int size, char type) {
```

Comment:
this function should be inline (see below comment to penalty_by_size)

Another thought: couldn't the case distinction on type be avoided by introducing three functions (I assume that the type is known or could be known at compile time in all calls to get_size_penalty).
(This could be done with templates as well:
template <char type>
PARAMTYPE size_penalties_class::get_size_penalty(int size);
)